ibis/expr/types/arrays.py

```diff
@@ -1106,15 +1115,38 @@ def array(values: Iterable[V]) -> ArrayValue:
     │ [3, 42, ... +1]      │
     └──────────────────────┘
 
-    >>> ibis.array([t.a, 42 + ibis.literal(5)])
-    ┏━━━━━━━━━━━━━━━━━━━━━━┓
-    ┃ Array()              ┃
-    ┡━━━━━━━━━━━━━━━━━━━━━━┩
-    │ array<int64>         │
-    ├──────────────────────┤
-    │ [1, 47]              │
-    │ [2, 47]              │
-    │ [3, 47]              │
-    └──────────────────────┘
+    >>> ibis.array([t.a, 42 + ibis.literal(5)], type="array<float>")
+    ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    ┃ Cast(Array(Array), array<float64>) ┃
+    ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
+    │ array<float64>                     │
+    ├────────────────────────────────────┤
+    │ [1.0, 47.0]                        │
+    │ [2.0, 47.0]                        │
+    │ [3.0, 47.0]                        │
+    └────────────────────────────────────┘
     """
-    return ops.Array(tuple(values)).to_expr()
+    type = dt.dtype(type) if type is not None else None
+    if type is not None and not isinstance(type, dt.Array):
+        raise ValidationError(f"dtype must be an array, got {type}")
+
+    if isinstance(values, ir.Value):
+        if type is not None:
+            return values.cast(type)
+        elif isinstance(values, ArrayValue):
+            return values
+        else:
+            raise ValidationError(
+                f"If no type passed, values must be an array, got {values.type()}"
+            )
+
+    if values is None:
+        if type is None:
+            raise ValidationError("If values is None/NULL, dtype must be provided")
+        return ir.null(type)
+
+    values = tuple(values)
+    if len(values) > 0 and type is not None:
+        return ops.Array(values).to_expr().cast(type)
```

Comment:
Can we avoid doing operations like `cast` here? Doing this makes assumptions about how conversion should work in the backend and casting may or may not be the right approach. Instead, we should handle this in the backends.

It'll probably be extra code to do that.

Reply:
Not sure I totally understand, like ops.Array should always have a dtype present, and the visit_Array() methods should always deal with it?

I wanted to support specifying a dtype in the ibis.array() API. To get there, I thought it would be better to split up the process into two stages: 1. create the value 2. cast to the right type. This leans better on the already existing casting implementations. But are you saying there are cases where this two-step process doesn't work, ie you lose some precision or something?